### PR TITLE
Becca Feedback

### PR DESCRIPTION
**Pull request by CheezItMan**
Hey Becca,

Below are the comprehension questions we give students students.  I'll put my comments directly in your code far below.

-- Chris

# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | 
What was a design decision you made that changed over time over the project? | 
What was a concept you gained clarity on, or a learning that you'd like to share? | 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+# class Room -- do we need a room class or just a reservation one?
+require 'date'
+require 'awesome_print'
+
+# https://ruby-doc.org/stdlib-2.4.1/libdoc/date/rdoc/Date.html
+
+# a single reservations
+class Reservation
+  # attribute accessor
+  attr_accessor :start_date, :end_date, :room_number, :total_cost, :reservation_id, :discount, :block_id
```
Comment (CheezItMan) [inline lib/hotel.rb:12]: Does `total_cost` need to be a variable, could it instead be a method, so that if the start/end dates change, the cost doesn't need to be manually updated?

Also do all of these need to be accessors?

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    date_key = start_date
+
+    # loops through dates
+    while date_key < end_date
+
+      # loop through room numbers
+      available_rooms.each do |room|
+        #remove reserved rooms from available rooms array
+        if @date_reservation_hash[date_key] == nil
+          available_rooms = available_rooms
```
Comment (CheezItMan) [inline lib/hotel.rb:113]: This statement doesn't seem to do anything, what you could do instead is

```ruby

unless @date_reservation_hash[date_key] || @date_reservation_hash[date_key].include?(room) == nil
  available_rooms.delete(room)
end
```

the `if` below is similar

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    @reservation_id = @start_date.to_time.to_i / (60 * 60 * 24) + room_number #ID is start date as integer plus room number
+    @discount = discount
+    @block_id = block_id
+  end
+end
+
+# the hotel system
+class HotelSystem
+  # attribute accessor
+  attr_accessor :room_numbers, :blocked_rooms, :date_reservation_hash, :id_reservation_hash, :blocks_hash, :date_block_hash
```
Comment (CheezItMan) [inline lib/hotel.rb:30]: I see you have several different data structures keeping track of reservations and blocks.  Could this duplication of data prove problematic if you had to make changes later?

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+  # updates blocks hash with info about block: [start_date, end_date, rooms, discount]
+  def create_block(start_year, start_month, start_day, end_year, end_month, end_day, block_rooms, cost, discount)
+    # dates
+    start_date = Date.new(start_year, start_month, start_day)
+    end_date = Date.new(end_year, end_month, end_day)
+
+    # create an id for the block which is (100000*num_days + (start_date +  first room number)
+    num_days = end_date - start_date
+    block_id = ((num_days*100000) + start_date.to_time.to_i / (60 * 60 * 24) + block_rooms[0]).to_i
```
Comment (CheezItMan) [inline lib/hotel.rb:140]: This is a lot of calculations for an ID.  Is it needed?

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    # check what rooms are available
+    available_rooms = self.get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)[0]
+
+    blocked_rooms = self.get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)[1]
```
Comment (CheezItMan) [inline lib/hotel.rb:145]: You end up calling `get_available_rooms` twice here.  Could this be done once and then you can pull out the two sub-arrays?

File: specs/spec_helper.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+require "minitest/autorun"
+require "minitest/reporters"
+
+Minitest::Reporters.use!
+Minitest::Reporters::SpecReporter.new
+
+require_relative "../lib/hotel.rb"
+
+# initialize a hotel system
+describe "HotelSystem" do
```
Comment (CheezItMan) [inline specs/spec_helper.rb:16]: Typically we create separate `filename_spec.rb` files in the `specs` folder and use `require_relative` to require the spec helper and run the tests with the `rake` command.  

I can walk you through this later when you start.

File: specs/spec_helper.rb [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    # 3 blocks
+    marriott.create_block(start_year, start_month, start_day, end_year, end_month, end_day, [1, 2, 3, 4], cost, 0.15)
+    marriott.create_block(start_year, start_month, start_day, end_year, end_month, end_day, [5, 6, 7, 8], cost, 0.15)
+    marriott.create_block(end_year, end_month, end_day, end_year+2, end_month+2, end_day+2, [5, 6], cost, 0.15)
+    marriott.reserve_blocked_room(318052, 5)
+    marriott.reserve_blocked_room(318052, 6)
+
+
+    #assert
+    # can't reserve room in a block
```
Comment (CheezItMan) [inline specs/spec_helper.rb:90]: Good work going through edge cases, normally we have students put them in separate `it` blocks so the test output is easier to read.

File: lib/hotel.rb [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    @date_block_hash = {} #date key, rooms reserved in block
+    @blocks_hash = {} # key = block ID, value = array start, end, room numbers, rooms_booked, discount
+  end
+
+  # make a reservation
+  # updates attributes: id_reservation_hash and date_reservation_hash
+  def make_reservation(start_year, start_month, start_day, end_year, end_month, end_day, room_number, cost, discount, block_id)
+
+    # check what rooms are not reserved
+    available_rooms = self.get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)[0]
```
Comment (CheezItMan) [inline lib/hotel.rb:47]: Do you need to call `self.get_available_rooms` twice?

Also this `make_reservation` method is getting pretty big, maybe breaking it up into helper methods  or pushing some functionality into the `Reservation` class.

**Review by CheezItMan — COMMENTED** [review]
@beccaelenzil 

Nice work, you hit the learning goals of the project and covered the essentials.  There are some minor structural things to change, like breaking your tests and classes into separate files, __very minor__, and we can review how to write Minitest specs as well.  

There are generally two ways students end up solving the overlapping reservations problem, and you hit on one of them by marking each date a room is reserved or blocked off.  The other way is comparing reservations and looking for overlapping dates.  This way is much simpler to code, but perhaps a bit less efficient.  It's also the most common student solution I think.  

There's definitely some messiness in the code, but for being self-taught in Ruby, this is quite good.  

Some of my feedback is very general as this is a design assignment with little in the way of right answers.  Let me know any questions you have and we can talk by email or chat.

File: lib/hotel.rb (hunk 1 above)
Comment (beccaelenzil) [reply lib/hotel.rb:12]: They definitely don't all need to be accessors. I can look back and be more precise with this.

**Comment by beccaelenzil** [comment]
Thanks Chris! Yes, I was definitely taking my best guess with minitest as this is something that I don't have much experience with. All of the rest of your comments make sense. 

I'm going to dive into the web part of the curriculum, but I'll certainly return to this.

Thanks again.
